### DISCUSSION TITLE
Cleaning on a per schema level?

### DISCUSSION THREAD

**Discussion by m1212e** [sinclairzx81/typebox#809] [Q&A] [answered]
Hi,

I'm using a generator to generate typebox schemas from other parts of my project. I was wondering if it is somehow possible to not only allow additional properties on a per schema level, but also to define the cleaning behavior in that regard:
For some schemas I want to apply a Value.Clean call beforehand. I simply don't care about additional fields gettings sent along, I just want them to be discarded. For other models I don't want this to happen and instead just throw an error. This configuraiton happens on a per schema basis and the info if this should happen or not is available at time of code generation. So I could for instance just set a hypothetical `{ cleanAdditionalFields: true }` on my generated objects. Currently I work around this by wrapping all in an additional object and checking beforehand if the clean should be applied. But it would be super cool if this was just supported right out of the (type)box since this would increase compatibility with the whole typebox ecosystem and other libraries that use typebox.

Did you every consider something like this, or is this even possible?

Thanks very much for you work!

**Comment by sinclairzx81** [comment, marked as answer]
@m1212e Hi

I'm not sure there's much I can do to make TypeBox optionally Clean internally (not without depending on library specific schema keywords, which I'd like to avoid), but you can implement this functionality externally.

```typescript
import { Type, TSchema } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

export function Clean(schema: TSchema, value: unknown): unknown {
  return schema.cleanAdditionalFields === true ? Value.Clean(schema, value) : value
}

const A = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number()
}, {
  cleanAdditionalFields: true
})

const B = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number()
})

console.log(Clean(A, { x: 1, y: 2, z: 3, w: 4 })) // { x: 1, y: 2, z: 3 }
console.log(Clean(B, { x: 1, y: 2, z: 3, w: 4 })) // { x: 1, y: 2, z: 3, w: 4 }
```
The Value.* operations are intended to be combined and rearranged in lots of different of ways, and TypeBox isn't very opinionated as to how these functions are ultimately used. Creating a custom Clean function with schema dependent rules applied is inline with general anticipated usage of these functions though.

Does this help?

**Reply by m1212e** [reply]
I see, thank you very much! This is kinda what I need, but sometimes I pass on the schema to an external library which then is using them to validate incoming requests. Would be cool to have automatic cleaning right there. But then I'll see if I can implement that somehow. Thanks for taking the time!
